includes/admin/class-wc-rest-woopay-session-controller.php

```diff
@@ -56,11 +56,7 @@ public function get_session_data( WP_REST_Request $request ): WP_REST_Response {
 		/**
 		 * @psalm-suppress UndefinedClass
 		 */
-		$response = WooPay_Session::get_init_session_request();
-		// This was needed as the preloaded requests were not honoring the cart token and so were empty carts.
-		// It would be ideal to get this to successfully preload the cart data so WooPay doesn't need to make
-		// a separate request to get the cart data.
-		unset( $response['preloaded_requests'] );
+		$response = WooPay_Session::get_init_session_request( null, null, null, $request );
```

Comment:
I am not certain but instead of calling the `get_init_session_request()` method here, should we create a separate (maybe a common) method?

At the moment, it looks like `get_init_session_request()` is:
* Expecting a [POST request](https://github.com/Automattic/woocommerce-payments/blob/5b861b3dab0d0b672359833c422e2f91c8253b7a/includes/woopay/class-woopay-session.php#L438).
* [Creating a new customer](https://github.com/Automattic/woocommerce-payments/blob/5b861b3dab0d0b672359833c422e2f91c8253b7a/includes/woopay/class-woopay-session.php#L407) each time.

This may add some complexity where we may need to explore the difference between an HTTP POST request from the FE versus an HTTP GET request from WooPay.

Reply:
>This may be the intended behavior, but I wanted to be certain that this was the case.

Nice catches! I'm not the certain that's the intended behavior for this request either, so I'm looping in @bborman22, who has most context on this particular flow, to confirm. If it's confirmed that this is not the intended behavior I will log a new issue capturing what you brought 👍 

>Expecting a [POST request](https://github.com/Automattic/woocommerce-payments/blob/5b861b3dab0d0b672359833c422e2f91c8253b7a/includes/woopay/class-woopay-session.php#L438).

Regarding the line you pointed out here, just wanted to mention that it looks like that part is "covered" in [this `TODO` comment](https://github.com/Automattic/woopay/blob/4e031439482d21212fdbd0d42dea0333b851027d/src/Session/RedirectSession.php#L114). However, I'm all for getting all of this logged in a single issue so we can address both issues together.

Now, coming back to the review itself, although I agree that the concerns you raised are valid and pertinent, I think we should be addressing them separately since these potential issues got here prior to this PR. The changes present in this PR are mainly addressing a performance issue of this flow.

Reply:
> should we create a separate (maybe a common) method?

I was debating this initially, and this PR seems to get us a tiny bit closer to that (abstracting out all the steps in the init session flow and then we can have different init functions that only run the steps they want). 

If I'm thinking of it correctly, I believe we would want to create a new customer. The idea is that by this point in the WooPay flow, we have only ever called `get_woopay_minimum_session_data` on the merchant store, which wouldn't have created a customer. This GET request from WooPay to merchant is to simulate the usual call to `get_init_session_request`, just later in the flow. The one thing to confirm is that `wp_get_current_user` is getting the proper user from the passed in Cart Token, so that the subsequent `get_customer_id_by_user_id` lookup succeeds or fails as expected.

The POST issue is a nice catch that I missed, but as @cesarcosta99 mentioned isn't a factor until that TODO is getting addressed, which would likely be a part of https://github.com/Automattic/woopay/issues/2648

Reply:
> Now, coming back to the review itself, although I agree that the concerns you raised are valid and pertinent, I think we should be addressing them separately since these potential issues got here prior to this PR. The changes present in this PR are mainly addressing a performance issue of this flow.

Agreed. These issues were not introduced in this PR and the changes on this PR improve the flow 👍

> The one thing to confirm is that `wp_get_current_user` is getting the proper user from the passed in Cart Token, so that the subsequent `get_customer_id_by_user_id` lookup succeeds or fails as expected.

This is a good call.

Apologies for the delay. This took a bit longer to investigate than I anticipated, but it looked important enough to do a deeper dive than normal.

It appears like `wp_get_current_user()` is not getting the correct user. That means that the session we retrieve from this request (`GET /woopay/session`) is not fully correct. The key data pieces that may cause issues are the `user_id`, `customer_id`, and `order_id`. There are a few others, but they're dependent on one of these (e.g. `session_nonce`, `blog_checkout_url`, `return_url`).

Because this issue is a higher priority and is related, we may be able to apply the following solution in this PR:
```diff
diff --git a/includes/admin/class-wc-rest-woopay-session-controller.php b/includes/admin/class-wc-rest-woopay-session-controller.php
index 655b9cb67..155058dc6 100644
--- a/includes/admin/class-wc-rest-woopay-session-controller.php
+++ b/includes/admin/class-wc-rest-woopay-session-controller.php
@@ -7,6 +7,7 @@
 
 defined( 'ABSPATH' ) || exit;
 
+use Automattic\WooCommerce\StoreApi\Utilities\JsonWebToken;
 use WCPay\WooPay\WooPay_Session;
 use Automattic\Jetpack\Connection\Rest_Authentication;
 
@@ -52,6 +53,13 @@ class WC_REST_WooPay_Session_Controller extends WP_REST_Controller {
 	 * @return WP_Error|WP_REST_Response
 	 */
 	public function get_session_data( WP_REST_Request $request ): WP_REST_Response {
+		// TODO: Ensure "cart_token" is passed in the request.
+		$token   = $request->get_header( 'cart_token' );
+		$payload = JsonWebToken::get_parts( $token )->payload;
+		$user_id = $payload->user_id ?? null;
+		// TODO: Ensure valid "user_id".
+		wp_set_current_user( $user_id );
+
 		// phpcs:ignore
 		/**
 		 * @psalm-suppress UndefinedClass
diff --git a/includes/woopay/class-woopay-session.php b/includes/woopay/class-woopay-session.php
index 391a22e38..4bb12bb8b 100644
--- a/includes/woopay/class-woopay-session.php
+++ b/includes/woopay/class-woopay-session.php
@@ -437,6 +437,10 @@ class WooPay_Session {
 
 		$email = ! empty( $_POST['email'] ) ? wc_clean( wp_unslash( $_POST['email'] ) ) : '';
 
+		if ( ! is_null( $woopay_request ) ) {
+			$order_id = $checkout_data['order_id'] ?? null;
+		}
+
 		$request = [
 			'wcpay_version'        => WCPAY_VERSION_NUMBER,
 			'user_id'              => $user->ID,
```

Reply:
> It appears like wp_get_current_user() is not getting the correct user.

Reply:
> It appears like wp_get_current_user() is not getting the correct user.

I verified that as well and can confirm that's happening. @lovo-h, thank you for going deeper and proposing a solution 🚀 

I applied the first part of your suggestion in https://github.com/Automattic/woocommerce-payments/pull/8645/commits/7e7ed4a39a630ad5774be434224ea161193d59c9. We talked on Slack about this and agreed we could validate the cart token in the `register_rest_route` call, but actually that's not feasible as a header is not a parameter we can manipulate in the `args`. With that said, I created a separate callback to validate that as you suggested. I also ensured that we only set the current user when we have a valid user ID.

The part about ensuring `$order_id` is set has been applied in https://github.com/Automattic/woocommerce-payments/pull/8645/commits/a1fcf877642d953232122058e1a72beae4168464.

I won't add extra testing instructions as you have context on the solution you proposed. But just to make things clear, ideally we should test the following as well:

1. When the user is logged in on WooPay and logged out of the merchant store we want to ensure the current user is not being set, a customer ID is being created, and the order ID is set before the request is dispatched.
2. When the user is logged in on WooPay and logged in on the merchant store we want to ensure the current user is set, the customer ID is retrieved (and not created), and order ID is set before the request is dispatched.